### PR TITLE
User maintenance time windows configuration

### PR DESCRIPTION
**Pull request by ppatierno**
### Type of change

_Select the type of your PR_

- Enhancement / new feature

### Description

This PR addresses #1109 adding the user configuration for maintenance time windows to describe when operations which start a rolling update (i.e. certificate renewal) can take place.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

**Comment by ppatierno** [comment]
@tombentley @scholzj a couple of comments on this PR for which I prefer to have a review.

First, I realized that checking the maintenance time window could be done just one time at the beginning of the reconciliation but then I started to think about long rolling update which could take time so that the subsequent operations could not fall in the window (i.e. a ZK rolling update starts because the window is OK but then it's long and the K rolling update doesn't start because the window is not satisfied anymore).
For this reason, I decided to check the maintenance window every time it's needed.

Second, I am thinking about tests, not sure how and where to put them.

**Comment by tombentley** [comment]
> For this reason, I decided to check the maintenance window every time it's needed.

This means the cluster state can be different on different pods, even in the absence of errors and strangeness. With the ZK rolling changes we won't be able to conclude the pod 0 has been rolled and pod N possibly not. I _think_ this is OK, but it's worth thinking carefully about.

> Second, I am thinking about tests, not sure how and where to put them.

Not sure what tests you mean, but personally:

* I would write a "unit" test using mocks.
* I would factor out an interface for supplying the "current" time, and use that in the test to check various boundary conditions.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 1]
```diff
@@ -354,7 +361,7 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
             return Future.succeededFuture(this);
         }
 
-        Future<ReconciliationState> getZookeeperState() {
+        Future<ReconciliationState> getZookeeperDescription() {
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Because it was the only one still having the "old" name :-)

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 2]
```diff
@@ -224,11 +228,14 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
         private ConfigMap topicOperatorMetricsAndLogsConfigMap = null;
         private ConfigMap userOperatorMetricsAndLogsConfigMap;
 
+        private List<String> maintenanceTimeWindows;
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Genuine question, what's the advantage of having `kafkaAssembly.getSpec().getMaintenanceTimeWindows()` inside a method, if every time we call that method it means to call always two methods on `kafkaAssembly`? I would see to call it one time and saving a reference to the returned list (as happens now) but maybe declaring `maintenanceTimeWindows` as a local variable in the `isMaintenanceTimeWindowsSatisfied` where it's used. It's also true that it means calling `kafkaAssembly.getSpec().getMaintenanceTimeWindows()` more time in a single rinconcile while it's enough doing that just one time, for this reason the member field. Putting this in a method seems to be just a way to "hide" what's happening underneath for having more clean code.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 3]
```diff
@@ -456,9 +463,15 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
...
         Future<ReconciliationState> zkRollingUpdate() {
-            return withVoid(zkSetOperations.maybeRollingUpdate(zkDiffs.resource(), pod ->
-                isPodToRestart(zkDiffs.resource(), pod, zkAncillaryCmChange, this.clusterCa)
-            ));
+            return withVoid(zkSetOperations.maybeRollingUpdate(zkDiffs.resource(), pod -> {
+                boolean isPodToRestart = isPodToRestart(zkDiffs.resource(), pod, zkAncillaryCmChange, this.clusterCa);
+                boolean isSatisfiedBy = true;
+                // it makes sense to check maintenance windows if pod restarting is needed
+                if (isPodToRestart) {
+                    isSatisfiedBy = isMaintenanceTimeWindowsSatisfied();
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Well, let me try to explain ...
The way we roll ZK or K and the EO are different as you know. We have our logic for ZK and K but we are relying on native Kuvernetes rolling update for EO (applying some patch on the resource).
It's true that `isPodToRestart` evaluates if a ZK or K pod needs a restart for some reasons and only, in that case, it makes sense to check if the maintenance time windows are satisfied otherwise it's a useless call because in any case, we are not going to roll the pod.
In case of EO we are just calling the reconcile leaving the native Kubernetes support to execute rolling if something changes. Before that we don't know if the desired deployment is different from current one (see this as a potential `isPodToRestart`) so the reason why I call in any case the check on the maintenance time windows (which is useful to get the right cert generation to put on the deployment). 
What you say means that we should get a way for knowing early that desired deployment is different or not from the current one but consider that the cert generation annotation is already always a difference, this is the reason why this method put the annotation in any case (but now I needed to put the right one, current or new cert genaration).

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 4]
```diff
@@ -1169,6 +1220,41 @@ private boolean isPodToRestart(StatefulSet ss, Pod pod, boolean isAncillaryCmCha
...
+                    }
+                }
+                return isSatisfiedBy;
+            } catch (ParseException e) {
+                e.printStackTrace();
+                return false;
+            }
+        }
+
+        private int getDeploymentCaCertGenaration(Deployment dep, Ca ca) {
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Yeah it's the `isPodCaCertUpToDate` method where a `Util.annotations(Pod)` method is used. I see that it gets a `HasMetadata`, is it safe to use the same method but casting the `PodTemplateSpec` to `HasMetadata`. Maybe yes. Wdyt? Said that we could factor out a common method handling the `PodTemplateSpec`.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: So I am a little bit confused now. While thanks to inlining by JIT compiler we don't have performance issues, we have them in this case because each method call means contacting the apiserver. So it seems saving in a variable seems to be better, having the call just one time. What am I missing or I did miss understand from what you said?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: `PodTemplateSpec` does not inherit `HasMetadata, so if I understand what you're proposing I think it wouldn't work. The `PodTemplateSpec` is the Java type of the thing at `StatefulSet.spec.template` and `Deployment.spec.template`, so you'd just need to call the getters to get the `spec.template`.

[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: So it means that we have to get rid to use the above `Util.annotations` method for getting annotations or writing a new one which gets a `PodTemplateSpec`.
It sounds really strange that something which has metadata actually doesn't implement `HasMetadata` interface. :-)

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 5]
```diff
@@ -911,9 +924,15 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
         }
 
         Future<ReconciliationState> kafkaRollingUpdate() {
-            return withVoid(kafkaSetOperations.maybeRollingUpdate(kafkaDiffs.resource(), pod ->
-                isPodToRestart(kafkaDiffs.resource(), pod, kafkaAncillaryCmChange, this.clusterCa, this.clientsCa)
-            ));
+            return withVoid(kafkaSetOperations.maybeRollingUpdate(kafkaDiffs.resource(), pod -> {
+                boolean isPodToRestart = isPodToRestart(kafkaDiffs.resource(), pod, kafkaAncillaryCmChange, this.clusterCa, this.clientsCa);
```
[the remark this answers is not in the document]
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Another reason to do this is that `isPodToRestart()` does detailed logging about why we're rolling. It would be confusing to the reader of the log to see all that only for the decision to roll to be overruled by the maint window. If we need the ability to selectively disable the maint window stuff that's just an `honourMaintWindow` parameter to `isPodToRestart()`.

[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Logging stuff is really a good point. The user could see messages like "Rolling pod .. due to ... " and then nothing happens. So it's worth moving checking windows inside.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: > we have them in this case because each method call means contacting the apiserver

You misunderstand. `kafkaAssembly` is already a field of `ReconciliationState`, so I'm saying there's no need to add this field as well. The value can be obtained from `kafkaAssembly.getSpec().getMaintenanceTimeWindows()` which doesn't involve the the apiserver at all. It's just some getters on a `Kafka` that's already been fetched from the apiserver. If you only need that value once just use `kafkaAssembly.getSpec().getMaintenanceTimeWindows()`. If you need it in multiple places (which I don't think you do) I'm saying you can put `kafkaAssembly.getSpec().getMaintenanceTimeWindows()` in a method to remove the duplicate code.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 6]
```diff
@@ -1169,6 +1220,41 @@ private boolean isPodToRestart(StatefulSet ss, Pod pod, boolean isAncillaryCmCha
...
+        private boolean isMaintenanceTimeWindowsSatisfied() {
+            try {
+                boolean isSatisfiedBy = maintenanceTimeWindows == null || maintenanceTimeWindows.isEmpty();
+                if (!isSatisfiedBy) {
+                    Date date = new Date();
+                    for (String cron : maintenanceTimeWindows) {
+                        CronExpression cronExpression = new CronExpression(cron);
+                        // the user defines the cron expression in "his" timezone but CO pod
+                        // can be running on a different one, so setting it on the cron expression
+                        cronExpression.setTimeZone(TimeZone.getDefault());
```
[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Maybe I was wrong I should have used `Calendar.getInstance().getTimeZone()` for setting the time zone on the cronexpression.
The idea is that the user write the cron expression in his timezone but of course the TZ on the pod could be different so I force the pod TZ to the cron expression so that it will be evaluate againg the right time and TZ on the pod (while the user still see the cron expression string in his timezone).
The `CronTest.testUserTimeZoneVsPodTimeZone` does something like this.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Ok then, but it seems that a method is needed because it's used 3 times in the same method : 

```
private boolean isMaintenanceTimeWindowsSatisfied() {
            try {
                boolean isSatisfiedBy = maintenanceTimeWindows == null || maintenanceTimeWindows.isEmpty();
                if (!isSatisfiedBy) {
                    Date date = new Date();
                    for (String cron : maintenanceTimeWindows) {
...
...

```

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Currently nothing knows what the user's TZ _is_ though; lacking that, what does the cron expression even _mean_? If I ask to you give me a wake up call at 0730 my time, but don't tell you where in the world I am, how can you possibly call me at the right time?

[the remark this answers is not in the document]
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: >  I pass the "* * 14-15 * * ?" cron expression to the system which applies the pod TZ so for example if it's in the "Pacific/Easter", the cron expression should be evaluated as "* * 8-9 * * ?".

How does that conversion work if you've not told anyone that you're in Italy? The pod doesn't know. The resource doesn't know. What does 'apply' mean? Does it mean that the cron expression is taken to be in that TZ, and it will convert it before comparing with the `Date` when deciding if the expression is satisfied? In which case you should be applying CET/CEST to the expression, not `Pacific/Easter`.

I'm happy to be wrong about this, but right now I just don't understand how it can possibly work without knowing that TZ the user was using when they wrote the expression.

[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: I got your point ...
So let's consider the example of my in Italy and I want maintenance every day from 14 to 15.
The pod system is in "Pacific/Easter" zone so that maintenance should main from 8 to 9 (6 hours behind me).
Imagine that I want to test it so I could:

- have a cron expression like  "* * 14-15 * * ?" and then set on the CronExpression object MY TZ so for example "Europe/Rome"
- making a Date like November 26, 8 AM with  "Pacific/Easter" as TZ

so something like this:

```java
public void testUserTimeZoneVsPodTimeZone() throws ParseException {
        // the "user" writes the cron expression in his timezone
        // every second, every minute, hour 14 to 15 every day --> from 14:00 to 15:59
        CronExpression cronExpression = new CronExpression("* * 14-15 * * ?");

        // the pod is running on a "Pacific/Easter" timezone data center, so cron expression needs the right timezone for evaluation
        cronExpression.setTimeZone(TimeZone.getTimeZone(ZoneId.of("Europe/Rome")));

        // it's really 20:00 in "Pacific/Easter" but 14:00 for "user"
        Date d = Date.from(LocalDateTime.of(2018, 11, 26, 8, 00, 0).atZone(ZoneId.of("Pacific/Easter")).toInstant());
        assertTrue(cronExpression.isSatisfiedBy(d));
    }
```

Does it sound reasonable to you?
It really means that we have to provide the user a way to specify the TZ in the maintenance windows property.

[the remark this answers is not in the document]
Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Saying that the user has to define the time windows in UTC time it's simpler even because we can avoid to add more information to the CRD so I would go for this one. It means that we have to set the "UTC" timezone to the `CronExpression`.

> But either way we need to know the TZ at the point we're evaluating the cron expr.

Why we have to know that? The `new Date()` which is used for validating the cron expression create current Date with the pod timezone but then the cron expression will evaluate that in terms of "UTC" as we set on it.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 7]
```diff
@@ -1162,17 +1203,65 @@ private boolean isPodToRestart(StatefulSet ss, Pod pod, boolean isAncillaryCmCha
...
+                        if (cronExpression.isSatisfiedBy(date)) {
+                            isSatisfiedBy = true;
+                            break;
+                        }
+                    }
+                }
+                return isSatisfiedBy;
+            } catch (ParseException e) {
+                log.warn("The provided maintenance time windows list contains a not valid cron expression");
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: We should include the invalid expression(s). It would be annoying to have, say 20 expressions and not be told which was the bad one.

**Comment by ppatierno** [comment]
@tombentley what do you think about the KAO update with the `Supplier<Date>` for better testing (which is coming)?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 8]
```diff
@@ -199,14 +200,14 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
         private final Kafka kafkaAssembly;
         private final Reconciliation reconciliation;
 
-        private ClusterCa clusterCa;
+        ClusterCa clusterCa; /* test */
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Since you're commenting on the (absence) of visibility modifier I think it's better to emphasize that by putting the comment in the same position as the (missing) visibility modifier:

```java
/*test*/ ClusterCa clusterCa;
```

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java [hunk 9]
```diff
@@ -914,9 +915,9 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
             return withKafkaDiff(kafkaSetOperations.reconcile(namespace, kafkaCluster.getName(), kafkaSs));
         }
 
-        Future<ReconciliationState> kafkaRollingUpdate() {
+        Future<ReconciliationState> kafkaRollingUpdate(Supplier<Date> dateSupplier) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Really using `Supplier` is being lazy. It has the signature we need, but the semantics of the interface (or rather, in that case, the semantics implied by its membership of `java.util.function`) are that it should be used in functional-style programming. That's not how we're using it. I think we should avoid being lazy and declare our own `Time` interface. 

I'm a fine one to lecture here, as I'm guilty of this sort of laziness all the time. Please pull be up on it in the future.

Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Tbh I don't see a real reason for writing a new interface from scratch when `Supplier` provides what we need. The semantics is about supplying a `Date` which is what we need. Why should we write a new one? I didn't get your point, sorry.

Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: Would you argue that you're never going to declare another interface having a single no args method, simply because `Supplier` exists, and you can use that? Because logically you _could_. Likewise if you just needed an interface with a single 1-arg method there's `Function`. You could do this, and your code would work, but it wouldn't make it any simpler to understand. For a start everywhere where you used a `Supplier` you'd have to document the semantics you expected for _that particular use site_. For example, in a crypto library a `Supplier<Long>` that returns prime numbers is a very different thing than a `Supplier<Long>` that returns random numbers.

Having a dedicated interface provides somewhere to declare what the required semantics are and they allow you to leverage type safety. 

You could compare the use of `Supplier<Date>` here with the use of `Predicate<Pod>` for the rolling. I feel that `Predicate<Pod>` is _just about_ OK, because what we need really does have precisely the semantics of a predicate on `Pods`. But a `Supplier`'s Javadoc is "Represents a supplier of results.", well that doesn't really describe quite how we're using it here, does it? We're not using it in a abstract way. Supplying the current time is quite a concrete thing, so a more concrete interface is appropriate.

Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: I'm not saying that it doesn't work, or is not meaningful. I am saying that using `Supplier`, `Function` and friends everywhere is a slippery slope to less type-safe code. An interface is a contract, and the contract needed here is more refined than merely the supply of any old random `Date`: It is supposed to be the _current_ date. The problem isn't so much the use of `Supplier` per-se, it's more that we're moving the down a slippery slope.

Comment (ppatierno) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java]: I don't see the slippery slope :-) Even `Supplier<Date>` defines a contract ... "providing a date" which is what we need and is also well defined, or?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MaintenanceTimeWindowsTest.java [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+
+        async.await();
+    }
+
+    private String getClusterCaGenerationPod(String podName) {
+        Pod pod = this.mockClient.pods().inNamespace(NAMESPACE).withName(podName).get();
+        return pod.getMetadata().getAnnotations().get(Ca.ANNO_STRIMZI_IO_CLUSTER_CA_CERT_GENERATION);
+    }
+
+    private void testZkRollingUpdate(List<String> maintenanceTimeWindows, Supplier<Date> dateSupplier,
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MaintenanceTimeWindowsTest.java]: Might be handy to document what case this is testing, since it's not so clear from the method name.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MaintenanceTimeWindowsTest.java [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+        };
+
+        StatefulSet z = this.mockClient.apps().statefulSets().inNamespace(NAMESPACE).withName(KafkaResources.zookeeperStatefulSetName(NAME)).get();
+        z.getSpec().getTemplate().getMetadata().getAnnotations().put(Ca.ANNO_STRIMZI_IO_CLUSTER_CA_CERT_GENERATION, "1");
+        this.mockClient.apps().statefulSets().inNamespace(NAMESPACE).withName(KafkaResources.zookeeperStatefulSetName(NAME)).patch(z);
+
+        this.reconciliationState.zkRollingUpdate(dateSupplier).setHandler(handler);
+    }
+
+    private void testKafkaRollingUpdate(List<String> maintenanceTimeWindows, Supplier<Date> dateSupplier,
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MaintenanceTimeWindowsTest.java]: Might be handy to document what case this is testing, since it's not so clear from the method name.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MaintenanceTimeWindowsTest.java [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+            }
+        };
+
+        StatefulSet k = this.mockClient.apps().statefulSets().inNamespace(NAMESPACE).withName(KafkaResources.kafkaStatefulSetName(NAME)).get();
+        k.getSpec().getTemplate().getMetadata().getAnnotations().put(Ca.ANNO_STRIMZI_IO_CLUSTER_CA_CERT_GENERATION, "1");
+        k.getSpec().getTemplate().getMetadata().getAnnotations().put(Ca.ANNO_STRIMZI_IO_CLIENTS_CA_CERT_GENERATION, "0");
+        this.mockClient.apps().statefulSets().inNamespace(NAMESPACE).withName(KafkaResources.kafkaStatefulSetName(NAME)).patch(k);
+
+        this.reconciliationState.kafkaRollingUpdate(dateSupplier).setHandler(handler);
+    }
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MaintenanceTimeWindowsTest.java]: Do we need tests that the EO and TO respect the time windows too?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MaintenanceTimeWindowsTest.java (hunk 11 above)
Comment (ppatierno) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MaintenanceTimeWindowsTest.java]: Actually, it's private and re-used by the real `@Test` methods. What kind of doc should I add here?

Comment (ppatierno) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MaintenanceTimeWindowsTest.java]: Well, it doesn't really do a complete reconciliation, so I would be for `doKafkaRollingUpdate` instead of the `test` prefix.

Comment (tombentley) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MaintenanceTimeWindowsTest.java]: Yeah, fine.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MaintenanceTimeWindowsTest.java [hunk 13]
```diff
@@ -347,9 +445,9 @@ public boolean certRenewed() {
             }
         };
 
-        StatefulSet z = this.mockClient.apps().statefulSets().inNamespace(NAMESPACE).withName(KafkaResources.zookeeperStatefulSetName(NAME)).get();
+        StatefulSet z = this.mockClient.apps().statefulSets().inNamespace(NAMESPACE).withName(ZookeeperCluster.zookeeperClusterName(NAME)).get();
         z.getSpec().getTemplate().getMetadata().getAnnotations().put(Ca.ANNO_STRIMZI_IO_CLUSTER_CA_CERT_GENERATION, "1");
-        this.mockClient.apps().statefulSets().inNamespace(NAMESPACE).withName(KafkaResources.zookeeperStatefulSetName(NAME)).patch(z);
+        this.mockClient.apps().statefulSets().inNamespace(NAMESPACE).withName(ZookeeperCluster.zookeeperClusterName(NAME)).patch(z);
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MaintenanceTimeWindowsTest.java]: Out of interest, why did you change this back to using the `ZookeeperCluster` method?

Comment (ppatierno) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/MaintenanceTimeWindowsTest.java]: What I got from the Stanislav review (about ZK rolling update) is that we are providing `KafkaResources` just for the api which can be used by other people and that internally we can still use the model ones. Did I get it wrong?
